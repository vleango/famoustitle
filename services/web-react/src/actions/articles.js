import axios from 'axios';

import { ROOT_API_URL } from './Base';

export const fetchList = (filters = {}) => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {

                let resource = "articles";
                if(filters['tag']) {
                    resource = `articles?tag=${filters['tag']}`
                } else if(filters['date']) {
                    resource = `articles?date=${filters['date']}`
                } else if(filters['match']) {
                    resource = `articles?match=${filters['match']}`
                }

                const response = await axios.get(`${ROOT_API_URL}/${resource}`);
                let data = { ...response.data, selected: filters };
                dispatch(list(data));
                resolve(data);
            }
            catch (error) {
                reject(error);
            }
        });
    }
};

export const fetchArticlesArchiveList = () => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {
                const response = await axios.get(`${ROOT_API_URL}/articles/archives`);
                let data = { ...response.data };
                dispatch(archives(data));
                resolve(data);
            } catch (error) {
                reject(error);
            }
        });
    }
};

export const createItem = (data) => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {
                const response = await axios.post(`${ROOT_API_URL}/articles`, data, authHeader(getState));
                resolve(response.body);
            }
            catch (error) {
                reject(error);
            }
        });
    }
};

export const fetchItem = (id) => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {
                const response = await axios.get(`${ROOT_API_URL}/articles/${id}`);
                resolve(response.body);
                dispatch(item({article: response.data}));
            }
            catch (error) {
                reject(error);
            }
        });
    };
};

export const updateItem = (id, data) => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {
                const response = await axios.post(`${ROOT_API_URL}/articles/${id}`, data, authHeader(getState));
                resolve(response.body);
            }
            catch (error) {
                console.log(error);
                reject(error);
            }
        });
    }
};

export const removeItem = (id) => {
    return async (dispatch, getState) => {
        return new Promise(async(resolve, reject) => {
            try {
                const response = await axios.delete(`${ROOT_API_URL}/articles/${id}`, authHeader(getState));
                resolve(response.body);
            }
            catch (error) {
                console.log(error);
                reject(error);
            }
        });
    }
};

export const list = (data) => ({
    type: 'ARTICLE_LIST',
    data
});

export const archives = (data) => ({
    type: 'ARTICLES_ARCHIVE_LIST',
    data
});

export const item = (data) => ({
    type: 'ARTICLE_ITEM',
    data
});

export const update = (data) => ({
    type: 'ARTICLE_UPDATE',
    data
});

const authHeader = (state) => {
    let token = "";
    if(state() && state().auth && state().auth.token) {
        token = state().auth.token;
    }
    return {
        headers: {"Authorization": `Bearer ${token}`}
    };
};
