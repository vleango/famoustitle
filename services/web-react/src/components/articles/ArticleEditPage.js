import React, { Component } from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { Button, Form, FormGroup, Input } from 'reactstrap';

import Header from '../shared/headers/Header';

export class ArticleEditPage extends Component {

  constructor(props) {
    super(props);
    this.state = {
      title: props.article.title,
      body: props.article.body
    };
  }

  onSubmitArticle = (e) => {
  }

  onInputChange = (e) => {
    const field = e.target.name;
    const value = e.target.value;
    this.setState(() => ({ [field]: value }));
  }

  render() {
    return (
      <div>
        <Header resourceTitle={this.props.article.title} />
        <div className="container">
            <Form onSubmit={this.onSubmitArticle}>
              <div className="row">
                <div className="col">
                  <FormGroup>
                    <Input type="text"
                      name="title"
                      value={this.state.title}
                      placeholder="Title"
                      onChange={this.onInputChange} />
                  </FormGroup>
                  <FormGroup>
                    <Input type="textarea"
                      rows="20"
                      name="body"
                      value={this.state.body}
                      placeholder="Add your article"
                      onChange={this.onInputChange} />
                  </FormGroup>
                </div>
                <div className="col">
                  <Button className="float-right" color="primary" size="lg">Save</Button>
                  <h2>{this.state.title}</h2>
                  <hr />
                  <ReactMarkdown source={this.state.body} />
                </div>
              </div>
            </Form>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
	return {
		article: state.articles.show
	};
};

export default connect(mapStateToProps, null)(ArticleEditPage);
