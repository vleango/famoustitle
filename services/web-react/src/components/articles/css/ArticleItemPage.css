.article-show-container ol li {
    font-size: 2.3rem;
    font-style: italic;
}

.hljs {
    line-height: 2rem;
    padding: 1rem !important;
}

.art-image {
    height: 450px;
    position: relative;
    color: white;
    transition: all 450ms;
}

.article-item-subtitle {
    font-size: 1.5rem;
}

.body-markdown img {
    max-width: 100%;
}

.body-markdown.spacing {
    line-height: 4.5rem;
    word-break: break-word;
}

.body-markdown h1 {
    margin-top: 6rem;
    margin-bottom: 1.5rem;
}

.body-markdown h2 {
    margin-top: 8rem;
    margin-bottom: 1.5rem;
}

.body-markdown h3 {
    margin-top: 6rem;
    margin-bottom: 1.5rem;
}

.nav-brand-text {
    font-weight: 700;
    font-family: "Source Sans Pro", Helvetica, Arial, sans-serif;
    font-size: 2.5rem;
    margin-left: 2rem;
}

.nav-title {
    font-weight: bolder;
    font-size: 2rem;
}

.modal-article-img-dialog {
    max-width: 70%;
}

.modal-img {
    width: 100%;
}

.pointer {
    cursor: pointer;
}

@media (min-width: 992px) {
    .article-show-container {
        max-width: 768px;
    }
}

@media (max-width: 576px) {
    .modal-article-img-dialog {
        max-width: 100%;
    }
}
