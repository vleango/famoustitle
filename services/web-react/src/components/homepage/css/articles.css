.article__title {
  font-size: 1.5rem;
}

.article__subtitle {
  list-style: none;
  padding: 0;
  margin: 0;
  opacity: .7;
  color: #000;
  font-weight: 600;
}

.article__subtitle > li {
  display: inline-block;
  margin-right: 30px;
  font-size: 14px;
}

.article__subtitle--date {
  text-decoration: none !important;
  font-size: 14px;
  font-weight: 600;
  color: inherit;
}

.article__subtitle--tags > a {
  margin-right: 0.7rem;
}

.article__subtitle--tag {
  text-decoration: none !important;
  font-weight: 600;
  padding: 0;
  margin-right: 0.7rem;
  color: inherit;
}

.article__subtitle--tag:hover {
  color: #0056b3;
}

.article__contents {
  display: flex;
  align-items: flex-start;
}

.article__contents--image-container {
  width: 25%;
  box-shadow: 1px 1px 1px 1px rgba(0, 0, 0, 0.3);
}

.article__contents--image {
  width: 100%;
}

.article__contents--body {
  width: 70%;
  margin-left: 1.5rem;
}

/* medium devices */
@media (max-width: 768px) {
  .article__contents--image-container {
    display: none;
  }

  .article__contents--body {
    width: 100%;
    margin: 0;
  }
}
