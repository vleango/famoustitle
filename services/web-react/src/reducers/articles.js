import { forEach } from 'lodash';

const articlesReducerDefaultState = {
    index: {
        pagination: {
            currentPage: 0,
            totalPages: 0
        },
        resources: [],
        archives: {},
        tags: [],
        selected: {}
    },
    show:  {
        resource: null,
        editable_id: null
    },
    edit:  { resource: null }
};

export default (state = articlesReducerDefaultState, action) => {
    switch(action.type) {
        case 'ARTICLE_LIST':
            let tags = action.data.tags.buckets.map((bucket) => {
                return bucket["key"];
            });
            return {
                ...state,
                index: {
                    ...state.index,
                    resources: action.data.articles,
                    tags: tags,
                    selected: action.data.selected
                }
            };
        case 'ARCHIVE_ARTICLES_LIST':
            let archives = {};
            forEach(action.data.archives.buckets, (bucket) => {
                archives[bucket["key_as_string"]] = bucket["doc_count"];
            });

            return {
                ...state,
                index: {
                    ...state.index,
                    archives: archives
                }
            };
        case 'ARTICLE_ITEM':
            return {
                ...state,
                show: {
                    ...state.show,
                    resource: action.data.article
                }
            };
        case 'ARTICLE_EDITABLE':
            return {
                ...state,
                show: {
                    ...state.show,
                    editable_id: action.data.editable_id
                }
            };
        case 'ARTICLE_UPDATE':
            return {
                ...state
            };
        default:
            return state;
    }
};
