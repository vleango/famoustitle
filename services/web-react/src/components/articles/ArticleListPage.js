import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Button, ListGroup, ListGroupItem, Media } from 'reactstrap';
import { truncate } from 'lodash';
import moment from 'moment';

import { fetchList } from '../../actions/articles';
import Header from '../shared/headers/Header';

export class ArticleListPage extends Component {

  componentDidMount() {
    this.props.fetchList && this.props.fetchList();
  }

  render() {
    return (
      <div>
        <Header />
        <div className="clearfix p-3">
          <Button tag={Link} to={`/articles/new`} className="float-right" color="primary">Add</Button>{' '}
        </div>
        {
          this.props.articles.map((article) => {
            return (
              <ListGroup key={article.id}>
                <ListGroupItem tag={Link} to={`/articles/${article.id}`} action>
                  <Media className="mb-5">
                    <Media>
                      <Media object src="http://via.placeholder.com/128x128" alt="Generic placeholder image">
                      </Media>
                    </Media>
                    <Media body className="ml-4">
                      <Media heading>
                        {article.title}
                      </Media>
                      <p>{ article.author }</p>
                      <p>{ moment(article.created_at).format('MM-DD-YYYY HH:mm') }</p>
                      {truncate(article.body, {
                        'length': 150
                      })}
                    </Media>
                  </Media>
                </ListGroupItem>
              </ListGroup>
            )
          })
        }
      </div>
    );
  }
}

const mapStateToProps = (state) => {
	return {
		articles: state.articles.index.articles
	};
};

const mapDispatchToProps = (dispatch) => ({
  fetchList: () => dispatch(fetchList())
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleListPage);
