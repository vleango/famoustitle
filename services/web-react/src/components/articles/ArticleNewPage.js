import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Form, FormGroup, Input } from 'reactstrap';
import { split, map, trim, uniq } from 'lodash';
import { toastInProgress, toastSuccess, toastFail } from '../shared/Toast';

import { createItem } from '../../actions/articles';
import Header from '../shared/headers/Header';

export class ArticleNewPage extends Component {

    constructor(props) {
        super(props);
        this.state = {
            title: "",
            body: "",
            tags: "",
            submitting: false,
            errorMsg: "",
            token: props.token
        };
    }

    onSubmitArticle = async (e) => {
        e && e.preventDefault();
        const { title, body } = this.state;
        if(title === "" || body === "") {
            this.setState({ errorMsg: "title or body is blank" });
            return;
        }

        const toastID = toastInProgress("Saving in progress...");

        try {
            this.setState({ submitting: true, errorMsg: "" });
            const rawTags = split(this.state.tags, ',');
            const trimmedTags = map(rawTags, (tag) => { return trim(tag).toLowerCase() });
            const tags = uniq(trimmedTags);
            const { title, body } = this.state;
            const article = { title, body, tags };
            await this.props.createItem({ token: this.state.token, article });
            toastSuccess("Save successful!", toastID);
            this.props.history.push('/');
        }
        catch(error) {
            this.setState({ submitting: false, errorMsg: "server error" });
            toastFail(e, toastID);
        }
    };

    onInputChange = (e) => {
        const field = e.target.name;
        const value = e.target.value;
        this.setState(() => ({ [field]: value }));
    };

    render() {
        return (
            <Fragment>
                <Header resourceTitle="Create New Article" />
                <div className="container">
                    <Form onSubmit={this.onSubmitArticle} autoComplete="off">
                        <FormGroup>
                            <Input type="text"
                                   name="title"
                                   value={this.state.title}
                                   placeholder="Title"
                                   onChange={this.onInputChange} />
                        </FormGroup>
                        <FormGroup>
                            <Input type="text"
                                   name="tags"
                                   value={this.state.tags}
                                   placeholder="Tag"
                                   onChange={this.onInputChange} />
                        </FormGroup>
                        <FormGroup>
                            <Input type="textarea" rows="20"
                                   name="body"
                                   value={this.state.body}
                                   placeholder="Add your article"
                                   onChange={this.onInputChange} />
                        </FormGroup>

                        { this.state.errorMsg && <p>{this.state.errorMsg}</p> }

                        <div className="clearfix">
                            <Button color="primary float-right" disabled={this.state.submitting} size="lg">Save</Button>
                        </div>
                    </Form>
                </div>
            </Fragment>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        token: state.auth.token
    };
};

const mapDispatchToProps = (dispatch) => ({
    createItem: async (data) => await dispatch(createItem(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleNewPage);
