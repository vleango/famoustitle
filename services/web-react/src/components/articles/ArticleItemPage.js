import React, {Component, Fragment} from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import ReactMarkdown from 'react-markdown';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { Button } from 'reactstrap';
import { Modal } from 'reactstrap';
import Spinner from '../shared/Spinner';
import {Helmet} from "react-helmet";

import { fetchItem, removeItem, itemEditable } from '../../actions/articles';
import './css/ArticleItemPage.css';

import faUser from "@fortawesome/fontawesome-free-solid/faUser";
import faCalendarAlt from "@fortawesome/fontawesome-free-solid/faCalendarAlt";
import faTag from "@fortawesome/fontawesome-free-solid/faTag";
import fontawesome from "@fortawesome/fontawesome/index";
import { toastInProgress, toastSuccess, toastFail } from '../shared/Toast';
fontawesome.library.add(faUser, faCalendarAlt, faTag);

export class ArticleItemPage extends Component {

    constructor(props) {
        super(props);
        this.state = {
            imgModal: false,
            article: null,
            errorMsg: "",
            editable_id: null,
            deleting: false
        };
    }

    async componentDidMount() {
        window.scrollTo(0, 0);

        if(this.props.itemEditable) {
            try {
                const response = await this.props.itemEditable(this.props.match.params.id);
                this.setState({
                    editable_id: response["id"]
                });
            } catch (error) {}
        }

        if(this.props.fetchItem) {
            try {
                await this.props.fetchItem(this.props.match.params.id);
            } catch(error) {
                this.setState({
                    errorMsg: error.toString()
                });
            }
        }
    }

    componentWillReceiveProps = (nextProps) => {
        if(nextProps.article) {
            this.setState({
                article: nextProps.article
            });
        }
    };

    displayInfo = () => {
        return (
            <div className="pt-3 pb-5">
                <FontAwesomeIcon className="mr-2" icon="user"/>
                <span className="mr-5">{this.state.article.author}</span>

                <FontAwesomeIcon className="mr-2" icon="calendar-alt"/>
                <span className="mr-5">{ moment(this.state.article.created_at).format('MM-DD-YYYY HH:mm') }</span>

                { this.state.article.tags && this.state.article.tags.length > 0 && (
                    <Fragment>
                        <FontAwesomeIcon className="mr-2" icon="tag"/>
                        {
                            this.state.article.tags.map((tag) => {
                                return [
                                    <Link key={tag} className="mr-2" to={`/?tag=${tag}`}>{tag}</Link>
                                ]
                            })
                        }
                    </Fragment>
                )}
            </div>
        );
    };

    imgModalToggle = (props) => {
        this.setState({
            modalData: {alt: props.target.alt, src: props.target.src},
            imgModal: !this.state.imgModal,
        });
    };

    LinkRenderer = (props) => {
        return <a href={props.href} rel="noopener noreferrer" target="_blank">{props.children}</a>
    };

    ImgRenderer = (props) => {
        return <img className={"pointer"} onClick={this.imgModalToggle} alt={props.alt} src={props.src}>{props.children}</img>
    };

    displayBody() {
        return (
            <div className={`body-markdown spacing`}>
                <ReactMarkdown source={ this.state.article.body } renderers={{link: this.LinkRenderer, image: this.ImgRenderer}}/>
            </div>
        )
    }

    onDeleteArticle = async () => {
        const toastID = toastInProgress("Deleting in progress...");
        try{
            this.setState({ deleting: true, errorMsg: "" });
            await this.props.removeItem({id: this.props.match.params.id });
            toastSuccess("Success!", toastID);
            this.setState({ submitting: false, errorMsg: "" });
            this.props.history.push('/');
        } catch(error) {
            let msg = "server error";
            if(error && error.response) {
                msg = error.response.statusText;
            }
            this.setState({ submitting: false, errorMsg: msg });
            toastFail(msg, toastID);
        }
    };

    render() {
        return (
            <div className="canvas">
                <Helmet>
                    {this.state.article && <title>{this.state.article.title} - FamousTitle.com</title>}
                </Helmet>

                <div className="container article-show-container pt-5 pb-5">
                    { this.state.editable_id && (
                        <div className="clearfix">
                            <Button disabled={this.state.deleting} onClick={this.onDeleteArticle} className="float-right" color="info">Delete</Button>
                            <Button tag={Link} to={`/articles/${this.state.editable_id}/edit`} className="float-right mr-4" color="info">Edit</Button>
                        </div>
                    ) }

                    {/* message for when loading article */}
                    { !this.state.article && <Spinner /> }
                    { this.state.errorMsg }

                    {/* after article loads */}
                    { this.state.article && (
                        <Fragment>
                            <h1>{this.state.article.title}</h1>
                            { this.displayInfo() }
                            { this.displayBody() }
                        </Fragment>
                    ) }

                    {/* Img Modal */}
                    <Modal isOpen={this.state.imgModal} toggle={this.imgModalToggle} centered={true} size={'article-img-dialog'}>
                        {this.state.modalData && (
                            <a rel="noopener noreferrer" onClick={this.imgModalToggle} target="_blank" href={this.state.modalData.src}>
                                <img className='modal-img' alt={this.state.modalData.alt} src={this.state.modalData.src} />
                            </a>
                        )}
                    </Modal>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        article: state.articles.show.resource
    };
};

const mapDispatchToProps = (dispatch) => ({
    fetchItem: async (data) => await dispatch(fetchItem(data)),
    itemEditable: async (data) => await dispatch(itemEditable(data)),
    removeItem: async (data) => await dispatch(removeItem(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleItemPage);
