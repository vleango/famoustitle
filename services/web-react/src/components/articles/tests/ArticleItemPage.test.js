import React from 'react';
import { Router, Route } from 'react-router-dom';
import { shallow } from 'enzyme';
import { Input } from 'reactstrap';
import ReactMarkdown from 'react-markdown';

import { ArticleItemPage } from '../ArticleItemPage';
import { article } from './fixtures/articles';

let wrapper;

describe('Components', () => {
  describe('ArticleItemPage', () => {

    beforeEach(() => {
      wrapper = shallow(<ArticleItemPage article={article} />);
    });

    describe('Snapshot', () => {
      it('should correctly render ArticleItemPage', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('isAuthenticated', () => {
      beforeEach(() => {
        wrapper = shallow(<ArticleItemPage article={article} isAuthenticated={true} />);
      });

      describe('EditMode', () => {
        describe('contains title', () => {
          it('shows title form', () => {
            const target = { article: article, editMode: ['title'] };
            wrapper.setState({ ...target });
            expect(wrapper.containsMatchingElement(<Input name="title"></Input>)).toBeTruthy();
          });
        });

        describe('contains body', () => {
          it('shows body form', () => {
            const target = { article: article, editMode: ['body'] };
            wrapper.setState({ ...target });
            expect(wrapper.containsMatchingElement(<Input type="textarea" name="body"></Input>)).toBeTruthy();
          });
        });

        describe('does not contain title', () => {
          it('does not show title form', () => {
            const target = { article: article, editMode: [''] };
            wrapper.setState({ ...target });
            expect(wrapper.containsMatchingElement(<Input name="title"></Input>)).toBeFalsy();
          });
        });

        describe('does not contain body', () => {
          it('does not show body form', () => {
            const target = { article: article, editMode: [''] };
            wrapper.setState({ ...target });
            expect(wrapper.containsMatchingElement(<Input type="textarea" name="body"></Input>)).toBeFalsy();
          });
        })
      });

      describe('onMouseOver', () => {
        it('sets the editModeClass', () => {
          wrapper.instance().onMouseOver();
          expect(wrapper.state('editModeClass')).toBe('outline');
        });
      });

      describe('onMouseLeave', () => {
        it('sets the editModeClass', () => {
          wrapper.instance().onMouseLeave();
          expect(wrapper.state('editModeClass')).toBe('');
        });
      });

      describe('onTextClicked', () => {
        it('sets editMode for the name in e', () => {
          const e = { currentTarget: { dataset: { name: 'hello' } }};
          wrapper.instance().onTextClicked(e);
          expect(wrapper.state('editMode')).toEqual(['hello']);
        })

        describe('uniqueness', () => {
          it('saves the editMode with uniquely', () => {
            wrapper.setState({editMode: ['hello']});
            const e = { currentTarget: { dataset: { name: 'hello' } }};
            wrapper.instance().onTextClicked(e);
            expect(wrapper.state('editMode')).toEqual(['hello']);
          });
        });
      });

      describe('onCancelClicked', () => {
        describe('editMode', () => {
          it('removes the field name from editMode', () => {
            wrapper.setState({editMode: ['title'], article: article});
            const e = { target: { dataset: { name: 'title' } }};
            wrapper.instance().onCancelClicked(e);
            expect(wrapper.state('editMode')).toEqual([]);
            expect(wrapper.state('title')).toEqual(article.title);
          });
        });
      });

      describe('onSavedClicked', () => {
        describe('empty field', () => {
          it('does not call updateItem', () => {
            const updateItem = jest.fn();
            wrapper = shallow(<ArticleItemPage article={article} updateItem={updateItem} />);
            const e = { target: { dataset: { name: 'title' } }};
            wrapper.instance().onSavedClicked(e);
            expect(updateItem).not.toHaveBeenCalled();
          });
        });

        describe('editMode', () => {
          it('removes the field name from editMode, call updateItem, and setState of editMode and article', () => {
            const updateItem = jest.fn();
            const match = {params: {id: 1}};
            wrapper = shallow(<ArticleItemPage article={article} updateItem={updateItem} match={match} />);
            wrapper.setState({editMode: ['title'], article: article, title: 'hello'});
            const e = { target: { dataset: { name: 'title' } }};
            wrapper.instance().onSavedClicked(e);
            expect(wrapper.state('editMode')).toEqual([]);
            expect(updateItem).toHaveBeenCalled();
            expect(wrapper.state('article').title).toEqual('hello');
          });
        });

        describe('onRemoveClicked', () => {
          describe('success', () => {
            it('calls removeItem', async () => {
              const removeItem = jest.fn(() => Promise.resolve());
              const match = {params: {id: 1}};
              wrapper = shallow(<ArticleItemPage article={article} removeItem={removeItem} match={match} history={[]} />);
              await wrapper.instance().onRemoveClicked();
              expect(removeItem).toHaveBeenCalled();
            });
          });

          describe('error', () => {
            it('resets submitting to false', async () => {
              const removeItem = jest.fn(() => Promise.reject());
              const match = {params: {id: 1}};
              wrapper = shallow(<ArticleItemPage article={article} removeItem={removeItem} match={match} history={[]} />);
              await wrapper.instance().onRemoveClicked();
              expect(removeItem).toHaveBeenCalled();
              expect(wrapper.state('submitting')).toBe(false);
            });
          });
        });

        describe('onSubmitChanges', () => {
          it('calls preventDefault on e', () => {
            const e = { preventDefault: jest.fn() }
            wrapper.instance().onSubmitChanges(e);
            expect(e.preventDefault).toHaveBeenCalled();
          });
        });

      });

      describe('onPreviewClicked', () => {
        it('sets previewMode for the element uniquely', () => {
          wrapper.setState({ previewMode: ['body'] });
          const e = { target: { dataset: { name: 'body' } } };
          wrapper.instance().onPreviewClicked(e);
          expect(wrapper.state('previewMode')).toEqual(['body']);
        });
      });

      describe('onPreviewTextClicked', () => {
        it('removes the element from previewMode', () => {
          wrapper.setState({ previewMode: ['body'] });
          const e = { currentTarget: { dataset: { name: 'body' } } };
          wrapper.instance().onPreviewTextClicked(e);
          expect(wrapper.state('previewMode')).toEqual([]);
        });
      });

      describe('onPreviewExitClicked', () => {
        it('removes the element from preview Mode', () => {
          wrapper.setState({ previewMode: ['body'] });
          const e = { currentTarget: { dataset: { name: 'body' } } };
          wrapper.instance().onPreviewTextClicked(e);
          expect(wrapper.state('previewMode')).toEqual([]);
        });
      });

      describe('displayBody', () => {
        describe('previewMode has body', () => {
          it('displays body in previewMode', () => {
            const displaySpy = jest.spyOn(wrapper.instance(), 'displayPreview');
            const displayForm = jest.spyOn(wrapper.instance(), 'displayForm');
            const displayMarkdown = jest.spyOn(wrapper.instance(), 'displayMarkdown');
            wrapper.setState({ previewMode: ['body'], editMode: [] });
            wrapper.instance().displayBody();
            expect(displaySpy).toHaveBeenCalled();
            expect(displayForm).not.toHaveBeenCalled();
            expect(displayMarkdown).not.toHaveBeenCalled();
          });
        });

        describe('previewMode does not have body but editMode does', () => {
          it('displays body in the Form', () => {
            const displaySpy = jest.spyOn(wrapper.instance(), 'displayPreview');
            const displayForm = jest.spyOn(wrapper.instance(), 'displayForm');
            const displayMarkdown = jest.spyOn(wrapper.instance(), 'displayMarkdown');
            wrapper.setState({ previewMode: [], editMode: ['body'] });
            wrapper.instance().displayBody();
            expect(displaySpy).not.toHaveBeenCalled();
            expect(displayForm).toHaveBeenCalled();
            expect(displayMarkdown).not.toHaveBeenCalled();
          });
        });

        describe('previewMode does not have body and editMode does not have body', () => {
          it('displays body in markDown', () => {
            const displaySpy = jest.spyOn(wrapper.instance(), 'displayPreview');
            const displayForm = jest.spyOn(wrapper.instance(), 'displayForm');
            const displayMarkdown = jest.spyOn(wrapper.instance(), 'displayMarkdown');
            wrapper.setState({ article: {body: 'test'}, previewMode: [], editMode: [] });
            wrapper.instance().displayBody();
            expect(displaySpy).not.toHaveBeenCalled();
            expect(displayForm).not.toHaveBeenCalled();
            expect(displayMarkdown).toHaveBeenCalled();
          });
        });
      });

    });

  });
});
