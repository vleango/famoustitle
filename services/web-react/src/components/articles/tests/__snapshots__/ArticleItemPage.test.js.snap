// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ArticleItemPage Snapshot should correctly render ArticleItemPage 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticleItemPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="container pt-5 pb-5"
>
        <p>
                Loading...
        </p>
</div>,
      "className": "canvas",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            Loading...
</p>,
          null,
        ],
        "className": "container pt-5 pb-5",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Loading...",
          },
          "ref": null,
          "rendered": "Loading...",
          "type": "p",
        },
        null,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container pt-5 pb-5"
>
          <p>
                    Loading...
          </p>
</div>,
        "className": "canvas",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              Loading...
</p>,
            null,
          ],
          "className": "container pt-5 pb-5",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Loading...",
            },
            "ref": null,
            "rendered": "Loading...",
            "type": "p",
          },
          null,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
